### DISCUSSION TITLE
DOXYGEN Precompiler Wrappers in SPI Master

### DISCUSSION THREAD

**Discussion by thestumbler** [modm-io/modm#1095] [General]
I was following some code while debugging SPI interactions, and ran into something that looked wrong at first glance. For example, in the file `src/modm/architecture/interface/spi_master.hpp`, a huge swath of function header declarations are wrapped with inside an `#ifdef __DOXYGEN__` block. Why can all these function declarations be ignored if you simply don't want to generate documentation?

Clearly this works and is intentional, as I see this kind of pattern in many different modules, some with `#ifdef` and some with `#ifndef`. These SPI functions all get defined in the appropriate platform driver files anyway, so is the architecture layer there just for documentation?

**Comment by salkinium** [comment]
> so is the architecture layer there just for documentation?

Mostly yes. We would need to rewrite this to use concepts, as that would actually check the interface properly. But we didn't have those back when we started these interfaces.
